### PR TITLE
Add Proto Models for DragAndDropSort Interaction

### LINKED ISSUE #1132: Add new interaction objects for drag & drop
Add new protos for the interaction objects needed for the drag & drop operation.

**Target PR date**: 18 May 2020
**Target completion date**: 19 May 2020

### PR DESCRIPTION
**Pull request by aggarwalpulkit596**
## Explanation

Fixes #1132 

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (vinitamurthi) [inline domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: I couldn't see what has changed here, is it just formatting? If so, have you run the kotlin lint check on this?

File: model/src/main/proto/interaction_object.proto
Comment (vinitamurthi) [inline model/src/main/proto/interaction_object.proto:22]: Can we just replace this with repeated StringList directly or do we want to keep it a separate interaction object type itself for other reasons? @BenHenning

File: model/src/main/proto/interaction_object.proto
Comment (vinitamurthi) [inline model/src/main/proto/interaction_object.proto]: Lets maybe call this set_of_html_strings instead of list?

**Review by vinitamurthi — CHANGES_REQUESTED** [review]

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: Yeah for this kind of thing can we add lint check on pre commit hooks just like oppe web so that whosoever is reviewing don't waste hir/her time in checking linting issues

File: model/src/main/proto/interaction_object.proto
Comment (aggarwalpulkit596) [reply model/src/main/proto/interaction_object.proto:22]: We can't directly use the repeated keyword it was written as a comment

File: model/src/main/proto/interaction_object.proto
Comment (aggarwalpulkit596) [reply model/src/main/proto/interaction_object.proto]: Done

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (vinitamurthi) [reply domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: Yeah we do have one , its not on pre commit right now but its on our continuous integration. But it looks like you are working on the domain module and we haven't enabled that yet. One easy way to do it yourself is to install ktlint and run ktlint -f 'path to your file'
It will format whatever it can and throw errors whenever it cant

File: model/src/main/proto/interaction_object.proto
Comment (vinitamurthi) [reply model/src/main/proto/interaction_object.proto:22]: Ah okay that makes sense, thanks!

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: okay thanks will do 👍

File: model/src/main/proto/interaction_object.proto
Comment (vinitamurthi) [reply model/src/main/proto/interaction_object.proto]: This doesn't seem to be done? What I meant is change
`repeated StringList list = 1` to 
`repeated StringList set_of_html_strings = 1`

Comment (aggarwalpulkit596) [reply model/src/main/proto/interaction_object.proto]: Was an internet issue

Comment (vinitamurthi) [reply model/src/main/proto/interaction_object.proto]: Ah okay got it , thanks! This now LGTM

**Review by vinitamurthi — APPROVED** [review]

File: model/src/main/proto/interaction_object.proto
Comment (BenHenning) [inline model/src/main/proto/interaction_object.proto]: Nit: please remove extra unneeded newline.

File: model/src/main/proto/interaction_object.proto
Comment (BenHenning) [inline model/src/main/proto/interaction_object.proto:64]: While I know this & the other new doc strings you added are just following existing patterns, can we maybe make this more useful? I'm thinking a link to the backend structure might be nice for reference.

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: This is probably a bit nonsensical (since we can't re-input the same answer string back into the UI since there are multiple components involved), but the output here might look something like:

``a, b, c, 1, 2, w, x, y, z``

For the structure:
```json
{
  setOfHtmlStrings: [
    ["a", "b", "c"],
    ["1", "2"],
    ["w", "x", "y", "z"]
  ]
}
```

Should we add some additional markers to delineate the individual lists that are contained within the set?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @aggarwalpulkit596! I left a few comments.

File: model/src/main/proto/interaction_object.proto
Comment (aggarwalpulkit596) [reply model/src/main/proto/interaction_object.proto:64]: We can just add a single link at the top of the file as all the models are declared in a single file

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: We can do that but in that case some additional changes might be required at the backend part i guess?

File: model/src/main/proto/interaction_object.proto
Comment (aggarwalpulkit596) [reply model/src/main/proto/interaction_object.proto]: Will do

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (BenHenning) [reply domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: I think we just need to change the toAnswerString method here. That’s not actually being used for anything currently besides logging, so we can return whatever we want to. No backend changes are needed I think.

File: model/src/main/proto/interaction_object.proto
Comment (BenHenning) [reply model/src/main/proto/interaction_object.proto:64]: That sounds good.

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (vinitamurthi) [inline domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: Just to make sure I am reading this correct. This would basically do:
`[a b c], [1 2], [w x y z]`
for 
```{
  setOfHtmlStrings: [
    ["a", "b", "c"],
    ["1", "2"],
    ["w", "x", "y", "z"]
  ]
}
```
right? 
Super tiny nit but can we keep the delimiter as "," instead of " " for it.htmlList? 
So it would be  `[a,b,c], [1,2], [w,x,y,z]`

**Review by vinitamurthi — APPROVED** [review]
LGTM! Just one nit

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: Output for the above [a b c, 1 2, w x y z]

Comment (vinitamurthi) [reply domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: Oops sorry I misread it! Yes you are right. How about we make it as 
`[a,b,c], [1,2], [w,x,y,z]`
That feels a lot easier to read

**Comment by aggarwalpulkit596** [comment]
@vinitamurthi @BenHenning updated although there is some issue with my keyboard and because of that an incomplete commit-msg was pushed along with an unwanted file should i make one more commit or can I just revert the last commit?

File: model/src/main/proto/interaction_object.proto
Comment (BenHenning) [inline model/src/main/proto/interaction_object.proto]: Please reference a specific commit hash instead of the develop branch to create a permalink. Otherwise if the file is moved or renamed in the future this link won’t work anymore.

**Comment by BenHenning** [comment]
Thanks @aggarwalpulkit596. Regarding the commit, I suggest just submitting another to fix it. 

Otherwise the PR LGTM. Will approve once the remaining comments are resolved.

File: .idea/gradle.xml
Comment (vinitamurthi) [inline .idea/gradle.xml]: Are you sure you have to modify this file? If so, add a newline at the end

File: model/src/main/proto/interaction_object.proto
Comment (aggarwalpulkit596) [reply model/src/main/proto/interaction_object.proto]: done

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: done

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: Nit: comma after ``,`` and before ``postfix``.

Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt]: One alternative approach to ensure we use the set's own answer string function:

```kotlin
return setOfHtmlStringsList.joinToString { "[${it.htmlList.toAnswerString()}]" }
```

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt:66]: I should have mentioned this before, but can you actually add a test for this? I don't think there's currently a test suite for the extensions file, but it seems worthwhile and hopefully straightforward to set up.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @aggarwalpulkit596! This LGTM, though I realized a bit late that we really ought to add a test for the answer string function even if it's trivial.

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt:66]: I can write the test for functions but how do i generate the Interaction Object? i mean for testing purpose i need a predefined Interaction Object

Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt:66]: @BenHenning I've tried something can you tell whether the approach is correct or not

File: domain/src/test/java/org/oppia/domain/util/InteractionObjectExtensionsTest.kt
Comment (aggarwalpulkit596) [inline domain/src/test/java/org/oppia/domain/util/InteractionObjectExtensionsTest.kt:19]: @BenHenning @vinitamurthi would this work?

File: domain/src/test/java/org/oppia/domain/util/InteractionObjectExtensionsTest.kt
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/util/InteractionObjectExtensionsTest.kt]: Since this isn’t retrieving data like a typical getter, we should name it something closer to what it’s doing, such as createHtmlStringList.

File: domain/src/test/java/org/oppia/domain/util/InteractionObjectExtensionsTest.kt
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/util/InteractionObjectExtensionsTest.kt]: We should include details about the setup of the object being tested in the test name, as well as the action being done. Consider:

testToAnswerStr_listOfSetsOfHtmlStrings_multipleLists_correctlyFormatsElements

File: domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt
Comment (BenHenning) [reply domain/src/main/java/org/oppia/domain/util/InteractionObjectExtensions.kt:66]: The test you added looks good to me. Just had a few minor comments.

**Review by BenHenning — APPROVED** [review]
Thanks @aggarwalpulkit596. Just a couple of nits, otherwise this LGTM.

File: domain/src/test/java/org/oppia/domain/util/InteractionObjectExtensionsTest.kt
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/util/InteractionObjectExtensionsTest.kt]: done

File: domain/src/test/java/org/oppia/domain/util/InteractionObjectExtensionsTest.kt
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/util/InteractionObjectExtensionsTest.kt]: done

**Review by vinitamurthi — APPROVED** [review]

File: .idea/gradle.xml
Comment (aggarwalpulkit596) [reply .idea/gradle.xml]: This is outdated file and I have  already reverted this

**Pull request merged**
